optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.io.jaxb;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import javax.xml.XMLConstants;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.bind.UnmarshallerHandler;
+import javax.xml.bind.ValidationEvent;
+import javax.xml.bind.util.ValidationEventCollector;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Source;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMResult;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+
+import org.optaplanner.core.impl.io.OptaPlannerXmlSerializationException;
+import org.xml.sax.Attributes;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.XMLFilter;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.XMLFilterImpl;
+
+public final class GenericJaxbIO<T> implements JaxbIO<T> {
+    private static final int DEFAULT_INDENTATION = 2;
+
+    private static final String ERR_MSG_WRITE = "Unable to write the %s to XML.";
+    private static final String ERR_MSG_READ = "Unable to read the (%s) from XML.";
```

Comment:
soft: I think putting error messages in a constant - if it's only used once - is a form of unneeded indirection, making the code harder to read (code is read 10 times more than it it written) because it reduces "vertical density" (that which belongs together is together).

In this case, the ERR_MSG_READ is used more than once, but that's often a sign that multiple different error causes are being reported as the same error, providing less information to the user. See examples below.

Reply:
These error message formatting are not consistent with our normal formatting "... parameterName (parameterValue) ..."
So
  "Unable to write the %s to XML."
=> "Unable to write rootClass (%s) to XML."

Reply:
- A rootClass isn't written to XML, it's marshalled to XML.
- We typically favor "Failed to" instead of "Unable to".
=> "Failed to marshall rootClass (%s) to XML."

Reply:
Only messages that repeat are defined as constants. However, I changed the message in a couple of places to make it more specific.

Thanks.